rastervision_core/rastervision/core/data/label/semantic_segmentation_labels.py

```diff
@@ -8,75 +10,279 @@
 
 
 class SemanticSegmentationLabels(Labels):
-    """A set of spatially referenced semantic segmentation labels."""
+    """Representation of Semantic Segmentation labels.
+
+    These labels can be either discrete or smooth.
+
+    If smooth, they store the scores for each class, for each pixel.
+    They also store the number of "hits" (initially equal to 0) for each pixel
+    that are used to divide the values for that pixel to get an average.
+    This allows one to do multiple updates to the same pixel
+    (the hits will be automatically incremented by 1 each time)
+    and have its final value be the average of all the updates.
+
+    If discrete, the labels are stored as a dict, mapping windows to label
+    arrays.
+
+    """
+
+    def __init__(self,
+                 smooth: bool = False,
+                 extent: Optional[Box] = None,
+                 num_classes: Optional[int] = None):
+        """Constructor.
```

Comment:
It seems like most of the methods follow the pattern of doing one thing if `smooth` and another if not. Maybe it would be cleaner if there were two subclasses of `SemanticSegmentationLabels` that implemented these different behaviors.

Reply:
On second thought maybe there should just be one class, but the implementation for the two modes should be unified so that for `smooth=False` it just uses a big array (with one fewer dimension) instead of the map to smaller arrays. The add method in this case would just overwrite instead of adding. Then it would make more sense to represent it as a single class, and seems like it would be simpler to understand.

Reply:
>but the implementation for the two modes should be unified so that for smooth=False it just uses a big array (with one fewer dimension) instead of the map to smaller arrays

This is what I had originally planned to do. But if there is a situation where the windows that we will be storing labels for make up only a small portion of the whole extent, creating an array for every pixel in the extent would not be very memory efficient. I also toyed with having a separate parameter to set the format as dict/array independent of the smooth/discrete param, but that ended up making the code too convoluted.